src/write_resume_data.cpp

```diff
@@ -0,0 +1,269 @@
+		{
+			// we need to save the whole merkle hash tree
+			// in order to resume
+			std::string& tree_str = ret["merkle tree"].string();
+			auto const& tree = atp.merkle_tree;
+			tree_str.resize(tree.size() * 20);
+			std::memcpy(&tree_str[0], &tree[0], tree.size() * 20);
+		}
+
+		if (!atp.unfinished_pieces.empty())
+		{
+			entry::list_type& up = ret["unfinished"].list();
+
+			// info for each unfinished piece
+			for (auto const& p : atp.unfinished_pieces)
+			{
+				entry piece_struct(entry::dictionary_t);
+
+				// the unfinished piece's index
+				piece_struct["piece"] = static_cast<int>(p.first);
+				std::string& bitmask = piece_struct["bitmask"].string();
+				for (auto bit : p.second)
+					bitmask.push_back(bit ? '1' : '0');
+				// push the struct onto the unfinished-piece list
+				up.push_back(std::move(piece_struct));
+			}
+		}
+
+		// save trackers
+		if (!atp.trackers.empty())
+		{
+			entry::list_type& tr_list = ret["trackers"].list();
+			tr_list.push_back(entry::list_type());
+			std::size_t tier = 0;
+			auto tier_it = atp.tracker_tiers.begin();
+			for (std::string const& tr : atp.trackers)
+			{
+				if (tier_it != atp.tracker_tiers.end())
+					tier = aux::clamp(std::size_t(*tier_it++), std::size_t{0}, std::size_t{1024});
+
+				if (tr_list.size() <= tier)
+					tr_list.resize(tier + 1);
+
+				tr_list[tier].list().push_back(tr);
+			}
+		}
+
+		// save web seeds
+		if (!atp.url_seeds.empty())
+		{
+			entry::list_type& url_list = ret["url-list"].list();
+			std::copy(atp.url_seeds.begin(), atp.url_seeds.end(), std::back_inserter(url_list));
+		}
+
+		if (!atp.http_seeds.empty())
+		{
+			entry::list_type& url_list = ret["httpseeds"].list();
+			std::copy(atp.http_seeds.begin(), atp.http_seeds.end(), std::back_inserter(url_list));
+		}
+
+		// write have bitmask
+		entry::string_type& pieces = ret["pieces"].string();
+		pieces.resize(aux::numeric_cast<std::size_t>(std::max(
+			atp.have_pieces.size(), atp.verified_pieces.size())));
+
+		std::size_t piece(0);
+		for (auto const bit : atp.have_pieces)
+		{
+			pieces[piece] = bit ? 1 : 0;
+			++piece;
+		}
+
+		piece = 0;
+		for (auto const bit : atp.verified_pieces)
+		{
+			pieces[piece] |= bit ? 2 : 0;
+			++piece;
+		}
+
+		// write renamed files
+		if (!atp.renamed_files.empty())
+		{
+			entry::list_type& fl = ret["mapped_files"].list();
+			for (auto const& ent : atp.renamed_files)
+			{
+				std::size_t const idx(static_cast<std::size_t>(static_cast<int>(ent.first)));
+				if (idx >= fl.size()) fl.resize(idx + 1);
+				fl[idx] = ent.second;
+			}
+		}
+
+		// write local peers
+		if (!atp.peers.empty())
+		{
+			std::back_insert_iterator<entry::string_type> ptr(ret["peers"].string());
+#if TORRENT_USE_IPV6
+			std::back_insert_iterator<entry::string_type> ptr6(ret["peers6"].string());
+#endif
+			for (auto const& p : atp.peers)
+			{
+#if TORRENT_USE_IPV6
+				if (p.address().is_v6())
+				{
+					write_address(p.address().to_v6(), ptr6);
```

Comment:
you mean to move the decision to write to ``ptr6`` or ``ptr`` into ``write_endpoint()``? (that seems like it could be a bit messy). Or do you mean to replace the two calls ``write_address()`` and ``write_uint16()`` with a single call to ``write_endpoint()``? Which seems reasonable to me. Assuming the function is inlined, that condition would be optimized out anyway.